src/sdk/python/tests/iotea/core/test_protocol_gateway.py

```diff
@@ -0,0 +1,251 @@
+        protocol_gateway.adapters[0].instance.publish.assert_called_once_with(test_topic, json.dumps(test_message,
+                                                                                                     separators=(
+                                                                                                         ',', ':')),
+                                                                              publish_options)
+
+
+@pytest.mark.asyncio
+# pylint: disable=redefined-outer-name
+async def test_publish_options(test_case, mocker, protocol_gateway):
+    test_topic = 'test'
+    test_message = 'test message'
+    mocker.patch('src.iotea.core.util.mqtt_client.MqttProtocolAdapter.publish', wraps=mock_publish)
+    # test publish with None pub options
+    await protocol_gateway.publish(test_topic, test_message, None)
+    test_case.assertEqual(1, protocol_gateway.adapters[0].instance.publish.call_count)
+    protocol_gateway.adapters[0].instance.publish.assert_called_once_with(test_topic, test_message, unittest.mock.ANY)
+    args, _ = protocol_gateway.adapters[0].instance.publish.call_args
+    # check that public options is not None
+    test_case.assertIsNotNone(args[2])
+
+    # test publish to platform adapter
+    publish_options = ProtocolGateway.create_publish_options(True, None)
+    await protocol_gateway.publish(test_topic, test_message, publish_options)
+    test_case.assertEqual(2, protocol_gateway.adapters[0].instance.publish.call_count)
+    protocol_gateway.adapters[0].instance.publish.assert_called_with(test_topic, test_message, publish_options)
+
+
+@pytest.mark.asyncio
+async def test_subscribe_options(test_case, mocker, protocol_gateway):
+    test_topic = 'test'
+
+    # pylint: disable=unused-argument
+    async def mock_subscribe(topic, cb, subscribe_options=None):
+        pass
+
+    mocker.patch('src.iotea.core.util.mqtt_client.MqttProtocolAdapter.subscribe', wraps=mock_subscribe)
+    # test with None sub options
+    await protocol_gateway.subscribe(test_topic, lambda: {}, subscribe_options=None)
+    test_case.assertEqual(1, protocol_gateway.adapters[0].instance.subscribe.call_count)
+    protocol_gateway.adapters[0].instance.subscribe.assert_called_with(test_topic, unittest.mock.ANY, unittest.mock.ANY)
+    args, _ = protocol_gateway.adapters[0].instance.subscribe.call_args
+    subscribe_options = args[2]
+    test_case.assertIsNotNone(subscribe_options)
+
+    # test with custom options: subscribe with platform protocol adapter
+    subscribe_options = ProtocolGateway.create_subscribe_options(True, None)
+    await protocol_gateway.subscribe(test_topic, lambda: {}, subscribe_options=subscribe_options)
+    test_case.assertEqual(2, protocol_gateway.adapters[0].instance.subscribe.call_count)
+    protocol_gateway.adapters[0].instance.subscribe.assert_called_with(test_topic, unittest.mock.ANY, subscribe_options)
+
+
+def test_validate_configuration():
```

Comment:
You should actually test, whether the exception you are expecting is the exception raised by comparing the exception messages

Reply:
Added message checks.